plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java

```diff
@@ -137,6 +198,67 @@ void addDependencies(List<Library> libraries) throws CoreException {
     }   
   }
 
+  /**
+   * Remove any dependencies that were required by the previously-selected libraries and that are
+   * not required by the currently selected libraries.
+   * 
+   * @param selectedLibraries the currently selected libraries
+   * @param previousLibraries the previously-selected libraries
+   */
+  @VisibleForTesting
+  static void removeUnusedDependencies(Element dependencies,
+      Collection<Library> selectedLibraries,
+      Collection<Library> previousLibraries) {
+
+    Set<String> selectedDependencies = new HashSet<>();
+    for (Library library : selectedLibraries) {
+      for (LibraryFile libraryFile : library.getDirectDependencies()) {
+        MavenCoordinates coordinates = libraryFile.getMavenCoordinates();
+        String encoded = coordinates.getGroupId() + "|" + coordinates.getArtifactId();
+        selectedDependencies.add(encoded);
+      }
+    }
+    Verify.verify(selectedDependencies.isEmpty() == selectedLibraries.isEmpty());
+
+    Map<String, Node> currentDependencies = new HashMap<>();
+    NodeList children = dependencies.getChildNodes();
+    for (int i = 0; i < children.getLength(); i++) {
+      Node node = children.item(i);
+      if (node.getNodeType() == Node.ELEMENT_NODE) {
+        Element dependency = (Element) node;
+        String groupId = getValue(dependency, "groupId");
+        String artifactId = getValue(dependency, "artifactId");
+        String encoded = groupId + "|" + artifactId;
+        currentDependencies.put(encoded, node);
+      }
+    }
+    Verify.verify(currentDependencies.isEmpty() == (dependencies.getChildNodes().getLength() == 0));
```

Comment:
do you really think this could happen? If so, we should throw a more explicit exception.